### ISSUE TITLE
Crash: IllegalStateException in VideoDecodeThread

### ISSUE THREAD

**Issue by jpage4500** [alexeyvasilyev/rtsp-client-android#16]
Hi,

I was able to use the demo app to test this RTSP stream perfectly: rtsp://wowzaec2demo.streamlock.net/vod/mp4:BigBuckBunny_115k.mov

But, when I try to test RTSP video from IP Webcam (https://play.google.com/store/apps/details?id=com.pas.webcam&hl=en_US&gl=US) - it's corrupted video.

I tested this many months ago on an older version of this library and it worked.. any ideas what could be wrong?

- notice the white in the video below.. it stays that way the whole time
<img src="https://user-images.githubusercontent.com/1895553/146569926-94a1b7ed-2828-44b6-8695-72cb34f4676f.png" width=50% height=50%>
<img src="https://user-images.githubusercontent.com/1895553/146570313-320bb4e4-607e-49ac-86b4-c49d8612db0c.png" width=50% height=50%>

- screenshots from IP Webcam
<img src="https://user-images.githubusercontent.com/1895553/146570237-72bf8c13-0e14-4944-8bc2-7a30446eab96.png" width=50% height=50%>
<img src="https://user-images.githubusercontent.com/1895553/146569938-b5c407fc-895d-4430-8eaf-e44800ca718e.png" width=50% height=50%>

I do see the following in the logs repeated over and over
```
12-17 10:43:17.575	com.alexvas.rtsp.demo	12889	15344	D	VideoDecodeThread	 No output from decoder available
12-17 10:43:17.579	com.alexvas.rtsp.demo	12889	15331	V	LiveFragment	 onRtspVideoNalUnitReceived(length=1856, timestamp=383254685)
12-17 10:43:17.580	com.alexvas.rtsp.demo	12889	15331	I	LiveFragment	 NALs (1): NAL_SLICE (1855 bytes)
12-17 10:43:17.586	com.alexvas.rtsp.demo	12889	15331	W	FrameQueue	 Cannot add frame, queue is full
```

I've also had the demo app crash several times.. this case it was clicking on the Logs tab
```
12-17 10:44:50.042		12889	15344	E	AndroidRuntime	 FATAL EXCEPTION: Thread-3
12-17 10:44:50.042		12889	15344	E	AndroidRuntime	 Process: com.alexvas.rtsp.demo, PID: 12889
12-17 10:44:50.042		12889	15344	E	AndroidRuntime	 android.media.MediaCodec$CodecException: Error 0x80000000
12-17 10:44:50.042		12889	15344	E	AndroidRuntime	 	at android.media.MediaCodec.native_stop(Native Method)
12-17 10:44:50.042		12889	15344	E	AndroidRuntime	 	at android.media.MediaCodec.stop(MediaCodec.java:2300)
12-17 10:44:50.042		12889	15344	E	AndroidRuntime	 	at com.alexvas.rtsp.demo.decode.VideoDecodeThread.run(VideoDecodeThread.kt:78)
```

**Comment by jpage4500** [comment]
@alexeyvasilyev - just wanted to add more info to this one. I noticed your last update and am using the same Video/Audio decoder files in my app.

The video was still only partially displayed/corrupted and the app crashed after a minute or 2 (log below)

I did however test the same URL in VLC and noticed the same corruption so I don't think it's related to this library anymore.. it must be an issue with IP Webcam. But, I wanted to keep this issue opened in case you can see why the crash happened. Please let me know if I can get you more info to help debug

FWIW: it should be easy to reproduce both the corruption and crash.. 
- IP Webcam: https://play.google.com/store/apps/details?id=com.pas.webcam&hl=en_US&gl=US
- start server and the URL should be something like this: `rtsp://192.168.0.227:8080/h264_ulaw.sdp`

```
12-20 18:27:49.432    com.jpage4500.hubitat.debug    27736    20556    E    ACodec       [OMX.qcom.video.decoder.avc] ERROR(0x80001009)
12-20 18:27:49.432    com.jpage4500.hubitat.debug    27736    20556    E    ACodec       signalError(omxError 0x80001009, internalError -2147483648)
12-20 18:27:49.432    com.jpage4500.hubitat.debug    27736    20555    E    MediaCodec     Codec reported err 0x80001009, actionCode 0, while in state 6
12-20 18:27:49.433    com.jpage4500.hubitat.debug    27736    20556    E    ACodec       signalError(omxError 0x80001001, internalError -2147483648)
12-20 18:27:49.433    com.jpage4500.hubitat.debug    27736    20555    D    SurfaceUtils     disconnecting from surface 0xb400007d6e2c0cf0, reason disconnectFromSurface
12-20 18:27:49.433    com.jpage4500.hubitat.debug    27736    20555    E    MediaCodec     Codec reported err 0x80001001, actionCode 0, while in state 0
12-20 18:27:49.434    com.jpage4500.hubitat.debug    27736    20705    D    VideoDecodeThread     No output from decoder available
12-20 18:27:49.435    com.jpage4500.hubitat.debug    27736    20554    W    System.err     java.lang.IllegalStateException
12-20 18:27:49.435    com.jpage4500.hubitat.debug    27736    20554    W    System.err         at android.media.MediaCodec.native_dequeueOutputBuffer(Native Method)
12-20 18:27:49.435    com.jpage4500.hubitat.debug    27736    20554    W    System.err         at android.media.MediaCodec.dequeueOutputBuffer(MediaCodec.java:3452)
12-20 18:27:49.435    com.jpage4500.hubitat.debug    27736    20554    W    System.err         at com.jpage4500.hubitat.utils.rtsp.VideoDecodeThread.run(VideoDecodeThread.kt:80)
12-20 18:27:49.437    com.jpage4500.hubitat.debug    27736    20554    E    AndroidRuntime     FATAL EXCEPTION: Thread-48
12-20 18:27:49.437    com.jpage4500.hubitat.debug    27736    20554    E    AndroidRuntime     Process: com.jpage4500.hubitat.debug, PID: 27736
12-20 18:27:49.437    com.jpage4500.hubitat.debug    27736    20554    E    AndroidRuntime     java.lang.IllegalStateException
12-20 18:27:49.437    com.jpage4500.hubitat.debug    27736    20554    E    AndroidRuntime         at android.media.MediaCodec.native_dequeueInputBuffer(Native Method)
12-20 18:27:49.437    com.jpage4500.hubitat.debug    27736    20554    E    AndroidRuntime         at android.media.MediaCodec.dequeueInputBuffer(MediaCodec.java:2855)
12-20 18:27:49.437    com.jpage4500.hubitat.debug    27736    20554    E    AndroidRuntime         at com.jpage4500.hubitat.utils.rtsp.VideoDecodeThread.run(VideoDecodeThread.kt:55)
12-20 18:27:49.445    com.jpage4500.hubitat.debug    27736    20705    D    VideoDecodeThread     No output from decoder available
12-20 18:27:49.456    com.jpage4500.hubitat.debug    27736    20705    D    VideoDecodeThread     No output from decoder available
12-20 18:27:49.459    com.jpage4500.hubitat.debug    27736    20707    E    ACodec       [OMX.qcom.video.decoder.avc] ERROR(0x80001009)
12-20 18:27:49.459    com.jpage4500.hubitat.debug    27736    20707    E    ACodec       signalError(omxError 0x80001009, internalError -2147483648)
12-20 18:27:49.459    com.jpage4500.hubitat.debug    27736    20706    E    MediaCodec     Codec reported err 0x80001009, actionCode 0, while in state 6
12-20 18:27:49.459    com.jpage4500.hubitat.debug    27736    20705    W    System.err     java.lang.IllegalStateException
12-20 18:27:49.460    com.jpage4500.hubitat.debug    27736    20705    W    System.err         at android.media.MediaCodec.native_dequeueOutputBuffer(Native Method)
12-20 18:27:49.460    com.jpage4500.hubitat.debug    27736    20705    W    System.err         at android.media.MediaCodec.dequeueOutputBuffer(MediaCodec.java:3452)
12-20 18:27:49.460    com.jpage4500.hubitat.debug    27736    20705    W    System.err         at com.jpage4500.hubitat.utils.rtsp.VideoDecodeThread.run(VideoDecodeThread.kt:80)
12-20 18:27:49.460    com.jpage4500.hubitat.debug    27736    20706    D    SurfaceUtils     disconnecting from surface 0xb400007d6e2b4090, reason disconnectFromSurface
12-20 18:27:49.460    com.jpage4500.hubitat.debug    27736    20707    E    ACodec       signalError(omxError 0x80001001, internalError -2147483648)
12-20 18:27:49.461    com.jpage4500.hubitat.debug    27736    20706    E    MediaCodec     Codec reported err 0x80001001, actionCode 0, while in state 0
12-20 18:27:49.462    com.jpage4500.hubitat.debug    27736    20705    E    AndroidRuntime     FATAL EXCEPTION: Thread-51
12-20 18:27:49.462    com.jpage4500.hubitat.debug    27736    20705    E    AndroidRuntime     Process: com.jpage4500.hubitat.debug, PID: 27736
12-20 18:27:49.462    com.jpage4500.hubitat.debug    27736    20705    E    AndroidRuntime     java.lang.IllegalStateException
12-20 18:27:49.462    com.jpage4500.hubitat.debug    27736    20705    E    AndroidRuntime         at android.media.MediaCodec.native_dequeueInputBuffer(Native Method)
12-20 18:27:49.462    com.jpage4500.hubitat.debug    27736    20705    E    AndroidRuntime         at android.media.MediaCodec.dequeueInputBuffer(MediaCodec.java:2855)
12-20 18:27:49.462    com.jpage4500.hubitat.debug    27736    20705    E    AndroidRuntime         at com.jpage4500.hubitat.utils.rtsp.VideoDecodeThread.run(VideoDecodeThread.kt:55)
12-20 18:27:49.551    com.jpage4500.hubitat.debug    27736    20705    I    Process      Sending signal. PID: 27736 SIG: 9

```

**Comment by alexeyvasilyev** [comment]
Tried to reproduce the problem with Pixel 4 XL running IP Webcam Pro and Demo app running on Pixel 6. Works like a charm. No any crashes.

I think the crash is related to your Android device.

**Comment by jpage4500** [comment]
Thanks for the reply @alexeyvasilyev; 

I am using the latest version in my app (RtspSurfaceView) and getting several crash reports coming in via Fabric/Crashlytics

```
Fatal Exception: java.lang.IllegalStateException
       at android.media.MediaCodec.native_dequeueInputBuffer(MediaCodec.java)
       at android.media.MediaCodec.dequeueInputBuffer(MediaCodec.java:983)
       at com.alexvas.rtsp.codec.VideoDecodeThread.run(VideoDecodeThread.kt:53)
```
and
```
Fatal Exception: java.lang.IllegalArgumentException: The surface has been released
       at android.media.MediaCodec.native_configure(MediaCodec.java)
       at android.media.MediaCodec.configure(MediaCodec.java:2023)
       at android.media.MediaCodec.configure(MediaCodec.java:1951)
       at com.alexvas.rtsp.codec.VideoDecodeThread.run(VideoDecodeThread.kt:43)
```
and
```
Fatal Exception: android.media.MediaCodec$CodecException: start failed
       at android.media.MediaCodec.native_start(MediaCodec.java)
       at android.media.MediaCodec.start(MediaCodec.java:612)
       at com.alexvas.rtsp.codec.VideoDecodeThread.run(VideoDecodeThread.kt:44)
```

So, these are reports coming in from other devices (not me). If I get more information such as how to reproduce this in your demo app I'll let you know

**Comment by alexeyvasilyev** [comment]
First. Line numbers are incorrect. Does not look like issues are from `RtspSurfaceView`
https://github.com/alexeyvasilyev/rtsp-client-android/blob/master/library-client-rtsp/src/main/java/com/alexvas/rtsp/codec/VideoDecodeThread.kt

Anyway, can you share a link (Google Play?) to your app to check?

The problem with MediaCodec based decoders that on crash, it shows a lot of useful information via logcat. So Fabric/Crashlytics is useless here. If you could get logcat log on crash, that would be more useful.

**Comment by jpage4500** [comment]
>> First. Line numbers are incorrect. Does not look like issues are from RtspSurfaceView

I added the first version you released and it wasn't available in a library yet.. that's probably why the source doesn't match up. I added it as a new module as-is so it probably was the same line numbers from the initial version.

I'll switch over to using the library version now that it's out. Also, I'll try to reproduce the crash myself and get a device log. The only RTSP video source I have is IP Webcam (app running on Android device) but I plan to get a camera that supports RTSP. That's where all of the crashes are likely coming from -- my app is a dashboard for user's smart home devices. It requires the Hubitat Hub though -- https://play.google.com/store/apps/details?id=com.jpage4500.hubitat

**Comment by jpage4500** [comment]
I finally got a camera that supports RTSP and am testing the demo app (latest master branch). I did make a few minor tweaks (PR https://github.com/alexeyvasilyev/rtsp-client-android/pull/23) if you're interested.

I'll try to do more stress testing and follow up on anything I find.

1 thing I did find out is that any app using RtspVideoView needs to be careful not to call `start` before the surface is created. In my case, I was calling init and start as soon as creating this view and was sometimes getting a crash because of it (one of the ones above)

I didn't update RtspVideoView but instead extended it and made sure the surface had been created before calling start.. however, in the future it could be ideal to just ensure nothing happens if start is called -- until the surface is created
